src/contract.ts

```diff
@@ -1,376 +1,479 @@
+     * @param program         Keypair for the account the program is located at
+     * @param storage         Keypair for the account the program's data is stored at
+     * @param space           Byte size to allocate for the storage account (this cannot be resized)
+     * @param options         @TODO: docs
+     *
+     * @return @TODO: docs
+     */
+    async deploy(
+        name: string,
+        constructorArgs: any[],
+        program: Signer,
+        storage: Signer,
+        space: number,
+        options?: ContractCallOptions
+    ): Promise<ContractCallResult> {
+        if (!program.publicKey.equals(this.program)) throw new Error('INVALID_PROGRAM_ACCOUNT'); // @FIXME: add error types
+        if (!storage.publicKey.equals(this.storage)) throw new Error('INVALID_STORAGE_ACCOUNT'); // @FIXME: add error types
+
+        const payer = options?.payer || this.payer;
+        if (!payer) throw new Error('MISSING_PAYER_ACCOUNT'); // @FIXME: add error types
+
+        const {
+            accounts = [],
+            writableAccounts = [],
+            programDerivedAddresses = [],
+            signers = [],
+            sender = payer.publicKey,
+            value = 0,
+            simulate = false,
+            confirmOptions = {
+                commitment: 'confirmed',
+                skipPreflight: false,
+                preflightCommitment: 'processed',
+            },
+        } = options ?? {};
+
+        const hash = keccak256(Buffer.from(name));
+        const seeds = programDerivedAddresses.map((pda) => pda.seed);
+        const input = this.interface.encodeDeploy(constructorArgs);
+
+        const data = Buffer.concat([
+            this.storage.toBuffer(), //                     storage @FIXME: these comments are kind of useless
+            sender.toBuffer(), //                           sender  @FIXME: better to explain why, not what
+            Buffer.from(numToPaddedHex(value), 'hex'), //   value
+            Buffer.from(hash.substr(2, 8), 'hex'), //       hash
+            encodeSeeds(seeds), //                          seeds
+            Buffer.from(input.replace('0x', ''), 'hex'), // input
+        ]);
+
+        // @FIXME: why are so many of these keys commented out?
+        const keys = [
+            // @FIXME: should all these PDAs really be writable?
+            // ...programDerivedAddresses.map((pubkey) => ({
+            //   pubkey,
+            //   isSigner: false,
+            //   isWritable: true,
+            // })),
+            {
+                pubkey: storage.publicKey,
+                isSigner: false,
+                isWritable: true,
+            },
+            // {
+            //   pubkey: SYSVAR_CLOCK_PUBKEY,
+            //   isSigner: false,
+            //   isWritable: false,
+            // },
+            // {
+            //   pubkey: PublicKey.default,
+            //   isSigner: false,
+            //   isWritable: false,
+            // },
```

Comment:
Given that the tests pass without all these accounts, do we need them?

Reply:
If you access `block.timestamp` or `block.slot` then you need `SYSVAR_CLOCK_PUBKEY`. Is there any reason not to include it?

Reply:
Yeah, I'm just wondering for these why they are commented out in the current implementation, and if it would still work if they were uncommented (I haven't tested)